File: Problem-060/solution.py
def prime_pair_sets():

    def is_prime(num):

        if num < 2:
            return False

        if num == 2:
            return True

        if num % 2 == 0:
            return False

        i = 3

        while i * i <= num:

            if num % i == 0:
                return False

            i += 2

        return True


    def compatible(a, b):

        ab = int(str(a) + str(b))
        ba = int(str(b) + str(a))

        return is_prime(ab) and is_prime(ba)


    primes = []

    for i in range(3, 10000, 2):


        if i != 5 and is_prime(i):
            primes.append(i)


    answer = float("inf")


    def search(selected, candidates, total):

        nonlocal answer
        if len(selected) == 5:

            answer = min(answer, total)
            return


        for i in range(len(candidates)):

            p = candidates[i]
            if total + p >= answer:
                break


            next_candidates = []

            
            for j in range(i + 1, len(candidates)):

                q = candidates[j]

                if compatible(p, q):
                    next_candidates.append(q)


            search(
                selected + [p],
                next_candidates,
                total + p
            )


    search([], primes, 0)

    return answer


print(prime_pair_sets())

File: Problem-060/solution.js
function primePairSets() {
    function isPrime(num) {
      if (num < 2) return false;
      if (num === 2) return true;
      if (num % 2 === 0) return false;
      for (let i = 3; i * i <= num; i += 2) {
        if (num % i === 0) {
          return false;
        }
      }
      return true;
    }
    function compatible(a, b) {
      let ab = Number(String(a) + String(b));
      let ba = Number(String(b) + String(a));
      return isPrime(ab) && isPrime(ba);
    }
    let primes = [];
    for (let i = 3; i < 10000; i += 2) {
      if (i !== 5 && isPrime(i)) {
        primes.push(i);
      }
    }
    let answer = Infinity;
    function search(set, candidates, sum) {
        if (set.length === 5) {
            if (sum < answer) {
                answer = sum;
            }
            return;
        }
        for (let i = 0; i < candidates.length; i++) {
            let p = candidates[i];
            if (sum + p >= answer) {
                break;
            }
            let nextCandidates = [];
            for (let j = i + 1; j < candidates.length; j++) {

                let q = candidates[j];

                if (compatible(p, q)) {
                    nextCandidates.push(q);
                }
            }
            search(
                [...set, p],
                nextCandidates,
                sum + p
            );
        }
    }
    search([], primes, 0);

    return answer;
}


console.log(primePairSets());
